### ISSUE TITLE
`\Magento\Customer\Api\CustomerRepositoryInterface::save` ignores `$passwordHash` for existing customers

### ISSUE THREAD

**Issue by Vinai** [magento/magento2#4256]
## Steps to reproduce
1. Install Magento from `develop` branch or version 2.0.4.
2. Create code to set a password for a customer using `CustomerRepositoryInterface::save`
3. Execute the code for an existing customer

``` php
$existingCustomer = $this->customerRepository->get($email);
$hash = $this->accountManagement->getPasswordHash($password);
$this->customerRepository->save($existingCustomer, $hash);
```
## Expected result

The customer is able to authenticate with the given password.
## Actual result

The given password hash is ignored, the old password still is active.

As a developer, this behavior is very surprising. Expecting a `save` method to update the password hash if it is supplied is reasonable.  

I think I can understand the reason (e.g. not exposing sensitive data via the web api), but having to debug the interface implementation and to understand how work around the behavior using non `@api` methods is time consuming and not a good experience. 

The workaround for folks ending up here using google:

``` php
$customerSecure = $this->customerRegistry->retrieveSecureData($customer->getId());
$customerSecure->setPasswordHash($hash);
$customerSecure->setRpToken(null);
$customerSecure->setRpTokenCreatedAt(null);
```

Besides being not easy to find, the downside of this workaround is it involves the `CustomerRegistry` and `CustomerSecure`, neither of which have an `@api` annotation.  
## Possible solution

It would be better to have the interface method names indicate what they do. 

**Option 1:**
Change the behavior of the `CustomerRepositoryInterface::save` method to update existing customers password hash if `$hash` was supplied.

**Option 2:**
Mark `save()` as `@deprecated`.
Add the more expressive methods `createWithPasswordHash($customer, $hash)` and `saveInsecureCustomerData($customer)`.
Add another method to set the customer password hash `updatePasswordHash($customer, $hash)` so all required functionality is exposed via stable service contract methods.

**Comment by Alkar778** [comment]
Your case not correct. Update customer password hash through `CustomerRepository::save` this is not secure way to update password.
In you case you should use `\Magento\Customer\Api\AccountManagementInterface::changePasswordById` or 
`\Magento\Customer\Api\AccountManagementInterface::resetPassword`
All this methods marked as `@api`

**Comment by Vinai** [comment]
@Alkar778, I think you closed the issue prematurely, as both the methods you mention do not cover the business use case I need.
`\Magento\Customer\Api\AccountManagementInterface::changePasswordById` requires knowledge of the old password, and `\Magento\Customer\Api\AccountManagementInterface::resetPassword` requires a reset token.

The required process is to be able to simply set a customer password to a given value using only `@api` methods.

Also, your answer doesn't reply to my statement that the design of the `CustomerRepositoryInterface::save` method is bad. Good code should not cause surprises or force a caller to look at the implementation of a method. The method signature should tell a caller enough so it can be used. This is doubly important for interfaces.  
With this in mind, the current state of the method should be considered broken.

**Comment by dlbnco** [comment]
Same problem here.

Trying to update customer password through REST API: 
`PUT V1/customers/{customerId}` 

But it won't update the `password_hash` column in the database.

**Comment by orlangur** [comment]
@dalbinaco you can find commits fixing problem in linked issue (they will be released in 2.2 only unless backported to 2.1).

**Comment by dlbnco** [comment]
@orlangur thank you

Checked on 2.2.0, it works!
